### PR TITLE
frontend: fix dialog closing behavior

### PR DESCRIPTION
**Pull request by misteroh**
### Description

Fix a bug with the  closing behavior of the Dialog frontend component: Closing the modal via ESC, the X button, or clicking the backdrop got the the modal stuck in an OFF state and unable to appear again.

The component now accepts an optional `onClose` callback. This callback can be used to set the `open` prop passed into `Dialog`. If this callback exists, the component will rely on the `open` prop alone to show/hide. Otherwise, the old functionality is preserved where the component will use its own internal `isOpen` state.  Warning: With the latter use case, closing the dialog with the Escape key, clicking on the backdrop, or using the "X" button will cause the modal to be stuck in the off state.

https://github.com/lyft/clutch/pull/674
https://github.com/lyft/clutch/pull/823

### Testing Performed
Storybook

### CODE REVIEW

File: frontend/packages/core/src/dialog.tsx
Comment (mcutalo88) [inline frontend/packages/core/src/dialog.tsx:72]: I dont know my typescript very well but is there a const for datatypes we could use over the string?

Comment (misteroh) [reply frontend/packages/core/src/dialog.tsx:72]: hmm... i guess I did define it in the `DialogProps` interface already so the typeof comparison doesn't have to be done... would `const onCloseHandlerExists = onClose ? true : false;` be better?

Comment (mcutalo88) [reply frontend/packages/core/src/dialog.tsx:72]: I will defer to you but that also looks good!

Comment (misteroh) [reply frontend/packages/core/src/dialog.tsx]: okay pushed out a new commit!

**Review by mcutalo88 — COMMENTED** [review]
Nice! Looks good, I'm going to defer final approval to @dschaller as he is the domain expert here.

File: frontend/packages/core/src/dialog.tsx
Comment (dschaller) [inline frontend/packages/core/src/dialog.tsx:67]: actually this completely makes sense since we already require state from the parent component. Let's change this to be required, remove the conditional and update the references.

**Comment by dschaller** [comment]
Awesome! Good catch :)

File: frontend/packages/core/src/dialog.tsx
Comment (misteroh) [reply frontend/packages/core/src/dialog.tsx]: done! can you please take a look and give me your thoughts on the `DialogCloseReasons` type? The Material Dialog unfortunately doesn't export the close reasons and instead just lists them in the JSdoc comment: 
image

File: frontend/packages/core/src/dialog.tsx
Comment (misteroh) [inline frontend/packages/core/src/dialog.tsx:68]: is `object` too generic? MuiDialog uses `{}` but our close button would pass a MouseEvent

**Comment by misteroh** [comment]
/lock

Can't merge until everything using Dialog passes onClose

**Comment by misteroh** [comment]
ready for review @dschaller !

File: frontend/packages/core/src/dialog.tsx
Comment (dschaller) [reply frontend/packages/core/src/dialog.tsx:68]: If we can make it more specific that would be great but it's not a blocker

**Review by dschaller — APPROVED** [review]

**Comment by dschaller** [comment]
Looks great! Thanks @misteroh

**Pull request merged**
